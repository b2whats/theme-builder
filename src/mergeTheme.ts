import { ThemeBuilder } from './ThemeBuilder'

const isObject = (value: any): value is object => value && value.constructor && value.constructor === Object
const isThemeBuilder = (value: any): value is ThemeBuilder<any> => value instanceof ThemeBuilder

function cloneUnlessOtherwiseSpecified(value: any) {
	return isObject(value)
		? mergeObject({}, value)
		: value
}

export function mergeObject(target: object, source: object): object {
  const result = {}

  Object.keys(target).forEach((key) => {
    result[key] = cloneUnlessOtherwiseSpecified(target[key])
  })

  Object.keys(source).forEach((key) => {
    const targetValue = target[key]
    const sourceValue = source[key]
    
    if (isThemeBuilder(targetValue) && isThemeBuilder(sourceValue)) {
      result[key] = targetValue.merge(sourceValue)
    } else if (isObject(targetValue) && isObject(sourceValue)) {
      result[key] = mergeObject(targetValue, sourceValue)
    } else {
      result[key] = cloneUnlessOtherwiseSpecified(sourceValue)
    }
  })

  return result
}

type UnaryFn<Arg, Return> = (arg: Arg) => Return

const weakMemoize = <Arg extends object, Return>(func: UnaryFn<Arg, Return>): UnaryFn<Arg, Return> => {
  let cache = new WeakMap()
  
  return arg => {
    if (cache.has(arg)) return cache.get(arg)

    let result = func(arg)
    cache.set(arg, result)
    
    return result
  }
}

let createCacheWithTheme = weakMemoize(outerTheme => {
  return weakMemoize(theme => {
    return mergeObject(outerTheme, theme)
  })
})

export function mergeTheme<T extends object>(target: T, source: object | void): T {
  if (!source) return target

  return createCacheWithTheme(target)(source) as T
}