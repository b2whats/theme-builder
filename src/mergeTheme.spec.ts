import { mergeObject, mergeTheme } from './mergeTheme'

export const target: object = {
  string: 'string',
  array: [1,2],
  number: 1,
  nestedObject: {
    object: {
      l: 1,
      m: 2,
      s: 3,
    },
    string: 'string',
    array: [1,2],
  },
  object: {
    l: 1,
    m: 2,
    s: 3,
  },
}

export const source: object = {
  string: 'string 2',
  array: [11],
  number: 11,
  nestedObject: {
    object: {
      l: 11,
    },
    string: 'string 2',
    array: [22],
    additionalNumber: 11
  },
  object: {
    s: 33,
  },
  additionalString: 'additionalString'
}


test('mergeObject', () => {
  const merge = mergeObject(target, source)

  expect(merge).toMatchSnapshot()
})

test('mergeObject - simple', () => {
  let firstArg = { one: 1, merge: 1 }
  let secondArg = { two: 2, merge: 2 }

  let firstResult = mergeTheme(firstArg, secondArg)
  let secondResult = mergeTheme(firstArg, secondArg)

  expect(firstResult).toBe(secondResult)
  expect(firstResult).toMatchSnapshot()

  let newArg = { two: 2, merge: 2 }
  let newResult = mergeTheme(firstArg, newArg)

  expect(newResult).not.toBe(firstResult)
  expect(newResult).toMatchSnapshot()
})