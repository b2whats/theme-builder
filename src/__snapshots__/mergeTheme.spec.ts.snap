// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mergeObject - call 1`] = `
Object {
  "merge": 2,
  "one": 1,
  "two": 2,
}
`;

exports[`mergeObject - simple 1`] = `
Object {
  "merge": 2,
  "one": 1,
  "two": 2,
}
`;

exports[`mergeObject - simple 2`] = `
Object {
  "merge": 2,
  "one": 1,
  "two": 2,
}
`;

exports[`mergeObject 1`] = `
Object {
  "additionalString": "additionalString",
  "array": Array [
    11,
  ],
  "nestedObject": Object {
    "additionalNumber": 11,
    "array": Array [
      22,
    ],
    "object": Object {
      "l": 11,
      "m": 2,
      "s": 3,
    },
    "string": "string 2",
  },
  "number": 11,
  "object": Object {
    "l": 1,
    "m": 2,
    "s": 33,
  },
  "string": "string 2",
}
`;
